feat: add variable float precision (#384)

dakhouya opened the pull request:
We can now change the number of decimal to display using shift left and shift right icons. The precision is going from 3 to 10 digit to keep it readable within the UI.

Shift left and shift right icons are only selectable when working with Float32 or Float64 buffer type.

brunoalr commented:
I didn't take a deep look at the code yet, but I see that Sonar caught 2 minor issues regarding `const`, can you fix them, please? [2 New issues](https://sonarcloud.io/project/issues?id=OpenImageDebugger_OpenImageDebugger&pullRequest=490&issueStatuses=OPEN,CONFIRMED&sinceLeakPeriod=true) - let me know if you have access to it, it should be public - otherwise I need to change some settings

dakhouya commented:
Yes I see them, I'll take a look!

brunoalr commented:
I played with it a little bit and I liked it. 
I found a bug regarding the size of rendered numbers with multi-channel images, even when they are int: 2-digit numbers are rendered bigger than 3-digit numbers.

Verified with `python3 out/OpenImageDebugger/oid.py --test`

Main:
![Screenshot 2025-01-07 at 21 01 07](https://github.com/user-attachments/assets/cb25ee25-9d43-46fe-9dac-c9982bb1ea89)

This branch:
![Screenshot 2025-01-07 at 21 02 04](https://github.com/user-attachments/assets/76850e41-4734-4aa4-98ea-d429eb04b32b)

Can you also fix it, please? :)

brunoalr commented:
One more request: it would be nice to show the current precision level somewhere in the screen - I don't know where tho, maybe the status screen?. Why: when the image is zoomed out and no numbers are shown it's not clear that the buttons changed anything.

dakhouya commented:
> One more request: it would be nice to show the current precision level somewhere in the screen - I don't know where tho, maybe the status screen?. Why: when the image is zoomed out and no numbers are shown it's not clear that the buttons changed anything.

dakhouya commented:
Sorry about the close, I miss click. I will probably take a look in the next days, Thanks for testing that, I only ran the testbench for my tests. 

For the precision, I agree that it would be a nice addition!

- [ ] Static analysis fix
- [ ] Resize bug
- [ ] Show current precision

dakhouya commented:
@brunoalr  I fixed the scaling and const issue, I'll check the UI in few days probably

By the time if you think about something else for the UI, let me know!

brunoalr commented:
> @brunoalr I fixed the scaling and const issue, I'll check the UI in few days probably
> 
> By the time if you think about something else for the UI, let me know!

Hi @dakhouya, I tested the lates commit and I still see the font scaling issue (now it's the time I promise myself I will add tests to the project haha)...

The feature seems complete enough for me. I didn't review the code itself yet.

dakhouya commented:
You are right @brunoalr, I'm fixing it. Probably a bad manipulation on my side while testing. 

For the fix, I reverted the way the scaling was done and reset text_pixel_scale to it's default value when the precision is changed. 

I push the change and rebased

dakhouya commented:
for a minimalist precision value display, it could look like that
it would be only present when working with float32 or float64

- float32x1
![image](https://github.com/user-attachments/assets/704a494e-2334-4eeb-a158-cee65e3f6231)

- uint8x3
![image](https://github.com/user-attachments/assets/1cd3bc95-38bb-4277-b65b-136bb94c9f6b)

dakhouya commented:
For the icons, I tried to search trough fontello since I have some trouble to integrate the custom svg within fontello - I end up with bad redering from figma to fontello.

Here what would already be available that might be better than what is currently integrated in the PR

![image](https://github.com/user-attachments/assets/1089ba29-3252-48dd-bde9-c00ec9b8a214)
![image](https://github.com/user-attachments/assets/4de08f4d-93c6-48bf-b66f-3bb0a1ad8f87)
![image](https://github.com/user-attachments/assets/1cf0b1be-db1e-4c01-8cef-4331318df1d0)

brunoalr commented on src/ui/main_window/main_window.ui:
```diff
@@ -378,6 +378,56 @@ QStatusBar::item { border: 0; };</string>
+                </size>
+               </property>
+               <property name="toolTip">
+                <string>Shift float precision left</string>
```
```suggestion
                <string>Decrease float precision</string>
```

brunoalr commented on src/ui/main_window/ui_events.cpp:
```diff
@@ -175,6 +175,73 @@ void MainWindow::link_views_toggle()
 
+void MainWindow::shift_precision_left()
+{
+    const auto shift_precision_left = [](Stage* stage) {
```
better avoid name shadowing
```suggestion
    const auto shift_precision_left_impl = [](Stage* stage) {
```

brunoalr commented on src/ui/main_window/ui_events.cpp:
```diff
@@ -175,6 +175,73 @@ void MainWindow::link_views_toggle()
+
+void MainWindow::shift_precision_right()
+{
+    const auto shift_precision_right = [](Stage* stage) {
```
```suggestion
    const auto shift_precision_right_impl = [](Stage* stage) {
```

brunoalr commented on src/ui/main_window/ui_events.cpp:
```diff
@@ -226,16 +293,17 @@ void MainWindow::rotate_90_ccw()
 void MainWindow::buffer_selected(QListWidgetItem* item)
 {
-    if (item == nullptr)
+    if (item == nullptr) {
```
nice catch, thanks :D

brunoalr commented:
tested it locally: font scaling is ok now :)
also reviewed. added some minor/nit comments, but overall, LGTM :)

brunoalr commented:
> For the icons, I tried to search trough fontello since I have some trouble to integrate the custom svg within fontello - I end up with bad redering from figma to fontello.
> 
> Here what would already be available that might be better than what is currently integrated in the PR
> 
> ![image](https://private-user-images.githubusercontent.com/1205218/407046989-1089ba29-3252-48dd-bde9-c00ec9b8a214.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg2ODUzMTYsIm5iZiI6MTczODY4NTAxNiwicGF0aCI6Ii8xMjA1MjE4LzQwNzA0Njk4OS0xMDg5YmEyOS0zMjUyLTQ4ZGQtYmRlOS1jMDBlYzliOGEyMTQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDIwNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAyMDRUMTYwMzM2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Zjk4MWY1ZTA1MGZkODgzZGM0YTU1ZWEyYjM4ZWU2ZWJmYTA0NDZjMzdhNjgzMzY5Njk3MDQxZmE1OGMzZmMwOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.Ilex7XJ24ZBbl1y3NpSoLw3xSczjSbFDUTvQBCD4xRQ) ![image](https://private-user-images.githubusercontent.com/1205218/407047028-4de08f4d-93c6-48bf-b66f-3bb0a1ad8f87.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg2ODUzMTYsIm5iZiI6MTczODY4NTAxNiwicGF0aCI6Ii8xMjA1MjE4LzQwNzA0NzAyOC00ZGUwOGY0ZC05M2M2LTQ4YmYtYjY2Zi0zYmIwYTFhZDhmODcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDIwNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAyMDRUMTYwMzM2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NjVjYTZlMjczZmFkMWY5YmUwMjM5YmFhMjg3ODI5ZjI3ZGQxNDg0MTgyMGY0ZjI0ZjQ1YTU2ZWZiOTU5MThiZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.0Rt7g8speWpECqEhDBEKni2jhwEK3eP0AGnSaIcgOTE) ![image](https://private-user-images.githubusercontent.com/1205218/407047078-1cf0b1be-db1e-4c01-8cef-4331318df1d0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg2ODUzMTYsIm5iZiI6MTczODY4NTAxNiwicGF0aCI6Ii8xMjA1MjE4LzQwNzA0NzA3OC0xY2YwYjFiZS1kYjFlLTRjMDEtOGNlZi00MzMxMzE4ZGYxZDAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDIwNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAyMDRUMTYwMzM2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZWQ2M2QyOGE3MWY4NmMzYWYzYjVjOTRlM2Q3NmJjODM4OGE0NmVkMmYwMzBiMTU0MzE1ZDVkNDRhNjIzYjQxMiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.WwUODpIDyxm42o6up5_qmNRo3hNXRm2p2bcKBHff9ZA)

for now, let's keep with what's already in the PR. thanks for researching :)

dakhouya commented:
@brunoalr changes are pushed in different commits
